### ISSUE TITLE
Refactor the style validator to use an off-the-shelf format like JSON Schema

### ISSUE THREAD

**Issue by lucaswoj** [mapbox/mapbox-gl-js#4159]
_From @lucaswoj on December 23, 2016 14:23_

Our ad-hoc JSON format that mixes validation logic & metadata can be tricky to work with.

Basing our validation on a tool like [JSON Schema](http://json-schema.org/documentation.html) would simplify and provide structure to our validation logic. It my also allow us to use some 3rd party validation, debugging, & documentation tools. 

_Copied from original issue: mapbox/mapbox-gl-style-spec#634_

**Comment by lucaswoj** [comment]
_From @jfirebaugh on December 23, 2016 16:33_

I think there are certain requirements that can't be captured in JSON Schema and will continue to require custom validation, but fewer than we have right now with our home grown schema.

**Comment by stevage** [comment]
Just a heads-up based on [my own experience](https://github.com/TerriaJS/terriajs-schema/tree/master/schema/2.0.0) with JSON schema, it can be hard to get good error messages from the validator when you have objects of this form:

```
{
   "type": "something",
    // every other attribute depends on what "something" is
}
```

...which is obviously true of quite a few objects in a style. It doesn't stop the validation working (although I did quite a bit of trial and error before I came up with a workable structure), you just sometimes get really meaningless error messages.


(If you want nightmares, have a look at [this](https://github.com/TerriaJS/generate-terriajs-schema/blob/master/generateSchema.js#L76)).

I also found that the standard validator in the JSON-schema library was a bit slow and needed to hack it, like to add some short-circuit validation failures. Incorporating the existing JSON schema library into Mapbox-GL-JS may not be a great idea - might be better to still write your own purpose-built validator (that happens to work on JSON schema).
